pip/pip-364.md

```diff
@@ -0,0 +1,476 @@
+and so on. When the score difference between the two paired brokers is greater than the threshold, the load will be evenly distributed
+between the two, which can solve the problem of slow speed.
+
+For example, if the broker rating of the current cluster is 20,30,52,70,80, we will pair 20 and 80, 30 and 70. As the difference between
+the two paired brokers is 80-20=60, 70-30=40, which are both greater than the threshold 40, the shedding strategy will be triggered.
+
+
+## handle load jitter with multiple hits threshold
+What about the historical weighting algorithm used in `ThresholdShedder`? It is used to solve the problem of load jitter, but previous
+analysis and experiments have shown that it can bring serious negative effects, so we can no longer use this method to solve the
+problem of load jitter.
+
+We mimic the way alarms are triggered: the threshold is triggered multiple times before the bundle unload is finally triggered. 
+For example, when the difference between a pair of brokers exceeds the threshold three times, load balancing is triggered.
+
+## high and low threshold
+In situations of cluster rolling restart or expansion, there is often a significant load difference between 
+different brokers, and we hope to complete load balancing more quickly. 
+
+Therefore, we introduce two thresholds:
+- loadBalancerAvgShedderLowThreshold, default value is 15
+- loadBalancerAvgShedderHighThreshold, default value is 40
+
+Two thresholds correspond to two continuous hit count requirements:
+- loadBalancerAvgShedderHitCountLowThreshold, default value is 8
+- loadBalancerAvgShedderHitCountHighThreshold, default value of 2
+
+When the difference in scores between two paired brokers exceeds the `loadBalancerAvgShedderLowThreshold` by 
+`loadBalancerAvgShedderHitCountLowThreshold` times, or exceeds the `loadBalancerAvgShedderHighThreshold` by
+`loadBalancerAvgShedderHitCountHighThreshold` times, a bundle unload is triggered. 
+For example, with the default value, if the score difference exceeds 15, it needs to be triggered 8 times continuously,
+and if the score difference exceeds 40, it needs to be triggered 2 times continuously.
+
+The larger the load difference between brokers, the smaller the number of times it takes to trigger bundle unloads, 
+which can adapt to scenarios such as cluster rolling restart or expansion.
+
+## placement strategy
+As mentioned earlier, `AvgShedder` bundles the shedding and placement strategies, and a bundle has already determined 
+its next owner broker based on the shedding strategy during shedding. But we not only use placement strategies after
+executing shedding, but also need to use placement strategies to assign bundles during cluster initialization, rolling
+restart, and broker shutdown. So how should we assign these bundles without shedding strategies?
+
+We use a hash allocation method: hash mapping (bundle name+random number) to broker. Hash mapping roughly conforms to 
```

Comment:
This is because a corner case was discovered during the testing process: assuming the Pulsar cluster has 4 nodes, the broker list is `[broker 1, broker2, broker3, broker4]`， After the cluster stabilizes, add a new broker 5, then close another broker 3, and the broker list will change to `[broker 1, broker2, broker5, broker4]`， **Broker5 replaces the position of broker 3**. As the total number of brokers in the cluster remains unchanged, most of the bundles on broker 3 are calculated using the hash algorithm to obtain the original index value (2). Therefore, **the vast majority of bundles on broker 3 will be transferred to broker 5**. 

As shown in the figure below, when turning off broker 3 (blue line), it is found that broker 5 (red line) has taken on all the load of broker 3. The message rate of broker 5 increased to broker 3's level, and the load (CPU usage) was almost the same as the other three brokers, while the load of the other three brokers has not changed much.
![image](https://github.com/apache/pulsar/assets/52550727/39e1597c-8528-4d6a-a2c3-e386d8ab413e)
![image](https://github.com/apache/pulsar/assets/52550727/f06748de-f1dc-4362-955f-dd2c69991ee5)

Reply:
Ok, you added `random number` to avoid this issue, but I wonder why don't we just use `random number` here, instead of using `bundle name+random number`? 

I checked the code, can we replace the code like below?

Before:
```
final long hashcode = hash(NamespaceBundle.getBundleRange(bundle.toString()) + random.nextInt());
final int index = (int) (Math.abs(hashcode) % sortedBrokers.size());
```
After:
```
final int index = Math.abs(random.nextInt() % sortedBrokers.size());
```

Reply:
Use random number only is ok too, i will update the doc and implementation.
```
final int index = Math.abs(hash(random.nextInt()) % sortedBrokers.size());
```